### PR TITLE
add sku option for sqldatabase

### LINKED ISSUE #126: azure_rm_sqldatabase - edition don't work
##### SUMMARY
This issue logically the same as https://github.com/ansible/ansible/issues/57965.
There is no way to specify `sku` for `azure_rm_sqldatabase` module. If I specify `edition` parameter - I have an exception.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
azure_rm_sqldatabase

##### ANSIBLE VERSION
```
ansible 2.9.9
  config file = /vagrant/ansible/ansible.cfg
  configured module search path = ['/home/vagrant/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vagrant/.local/lib/python3.6/site-packages/ansible
  executable location = /home/vagrant/.local/bin/ansible
  python version = 3.6.8 (default, Apr  2 2020, 13:34:55) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

##### CONFIGURATION
```
Not related.
```

##### OS / ENVIRONMENT
CentOS 7 + Python 3.6.8

```
$ pip3 list | grep azure
azure-cli-core (2.0.35)
azure-cli-nspkg (3.0.2)
azure-common (1.1.11)
azure-graphrbac (0.40.0)
azure-keyvault (1.0.0a1)
azure-mgmt-authorization (0.51.1)
azure-mgmt-automation (0.1.1)
azure-mgmt-batch (5.0.1)
azure-mgmt-cdn (3.0.0)
azure-mgmt-compute (4.4.0)
azure-mgmt-containerinstance (1.4.0)
azure-mgmt-containerregistry (2.0.0)
azure-mgmt-containerservice (4.4.0)
azure-mgmt-cosmosdb (0.5.2)
azure-mgmt-devtestlabs (3.0.0)
azure-mgmt-dns (2.1.0)
azure-mgmt-hdinsight (0.1.0)
azure-mgmt-iothub (0.7.0)
azure-mgmt-keyvault (1.1.0)
azure-mgmt-loganalytics (0.2.0)
azure-mgmt-marketplaceordering (0.1.0)
azure-mgmt-monitor (0.5.2)
azure-mgmt-network (2.3.0)
azure-mgmt-nspkg (2.0.0)
azure-mgmt-rdbms (1.4.1)
azure-mgmt-redis (5.0.0)
azure-mgmt-resource (2.1.0)
azure-mgmt-servicebus (0.5.3)
azure-mgmt-sql (0.10.0)
azure-mgmt-storage (3.1.0)
azure-mgmt-trafficmanager (0.50.0)
azure-mgmt-web (0.41.0)
azure-nspkg (2.0.0)
azure-storage (0.35.1)
msrestazure (0.5.0)
```

##### STEPS TO REPRODUCE
```yaml
- name: Create sql server
  azure_rm_sqlserver:
    resource_group: myrg
    name: myserver
    admin_username: myadmin
    admin_password: mypassword

- name: Create db
  azure_rm_sqldatabase:
    edition: basic
    resource_group: myrg
    server_name: myserver
    name: mydatabase
```

##### EXPECTED RESULTS
The database should be created with Basic SKU.

##### ACTUAL RESULTS
There is an error:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'edition'
failed: [localhost] ("module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 102, in <module>\n  File \"<stdin>\", line 94, in _ansiballz_main\n  File \"<stdin>\", line 40, in invoke_modul
e\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_s
pec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_azure_rm_sqldatabase_payload_vis6a89q/ansible_azure_rm_sqldatabase_payload.zip/ansibl
e/modules/cloud/azure/azure_rm_sqldatabase.py\", line 514, in <module>\n  File \"/tmp/ansible_azure_rm_sqldatabase_payload_vis6a89q/ansible_azure_rm_sqldatabase_payload.zip/ansible/modules/cloud/azure/azure_rm_sqldatabase.py\", li
ne 510, in main\n  File \"/tmp/ansible_azure_rm_sqldatabase_payload_vis6a89q/ansible_azure_rm_sqldatabase_payload.zip/ansible/modules/cloud/azure/azure_rm_sqldatabase.py\", line 314, in __init__\n  File \"/tmp/ansible_azure_rm_sql
database_payload_vis6a89q/ansible_azure_rm_sqldatabase_payload.zip/ansible/module_utils/azure_rm_common.py\", line 348, in __init__\n  File \"/tmp/ansible_azure_rm_sqldatabase_payload_vis6a89q/ansible_azure_rm_sqldatabase_payload.
zip/ansible/modules/cloud/azure/azure_rm_sqldatabase.py\", line 387, in exec_module\nKeyError: 'edition'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

I think the error in the line https://github.com/ansible-collections/azure/blob/dev/plugins/modules/azure_rm_sqldatabase.py#L387 in `old_response['edition']` part.

### LINKED ISSUE #153: azure_rm_sqldatabase lacks critical functionality
##### SUMMARY

Currently we are: 
    - Unable to specify a SKU
    - Unable to deploy vCores
    - Unable to create an elastic pool
    - Creating elastic pools using azure_rm_resource works, but when we want to add new databases to pool, we have to rerun the deployment
    - Unable to utilize hyperscale

We would like all of these capabilities added. 
##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

azure_rm_sqldatabase
##### ADDITIONAL INFORMATION

Should be self explanatory from above. These are all values needed to configure a sql database, however there's no documented method of doing so.

### PR DESCRIPTION
**Pull request by guyomog78**
##### SUMMARY
Change the non functional **edition** option to a functional *sku* option.

Partial Fixes #153 *Unable to specify a SKU*
Fixes #126 
##### ISSUE TYPE
- Feature Pull Request
- Bugfix Pull Request

##### COMPONENT NAME
azure_rm_sqldatabase.py
azure_rm_sqldatabase_info.py

##### ADDITIONAL INFORMATION

The Sku option use the Python *Sku* class from azure.mgmt.sql.models to be able to push valide values.
The documentation is updated with the new SKU call option
The internal test to detect change is updated (replace **edition** by **sku**)

The sku entrie is then identical as the azure_rm_sqldatabase_info report

**example call before change**
```
  - name: Create (or update) SQL Database with SKU
    azure_rm_sqldatabase:
      resource_group: myResourceGroup
      server_name: sqlcrudtest-5961
      name: testdb
      location: eastus
      edition: standard
```

**example call after change**
```
  - name: Create (or update) SQL Database with SKU
    azure_rm_sqldatabase:
      resource_group: myResourceGroup
      server_name: sqlcrudtest-5961
      name: testdb
      location: eastus
      sku:
        name: S0
```
or with more options
**example call after change**
```
  - name: Create (or update) SQL Database with SKU
    azure_rm_sqldatabase:
      resource_group: myResourceGroup
      server_name: sqlcrudtest-5961
      name: testdb
      location: eastus
      sku:
        name: S0
        tier: Standard
        capacity: 50
```

**Attention** All option under sku should be coherent with the name value else an error is reported

### CODE REVIEW

File: plugins/modules/azure_rm_sqldatabase.py [hunk 1]
```diff
@@ -77,23 +77,23 @@
         description:
             - Required if I(create_mode=restore_long_term_retention_backup), then this value is required.
             - Specifies the resource ID of the recovery point to restore from.
-    edition:
+    sku:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_sqldatabase.py:96]: Argument 'size' and 'family' found in sku is listed in the argument_spec, but not documented in the module documentation

File: plugins/modules/azure_rm_sqldatabase.py [hunk 2]
```diff
@@ -334,8 +343,9 @@ def exec_module(self, **kwargs):
                     self.parameters["restore_point_in_time"] = kwargs[key]
                 elif key == "recovery_services_recovery_point_resource_id":
                     self.parameters["recovery_services_recovery_point_resource_id"] = kwargs[key]
-                elif key == "edition":
-                    self.parameters["edition"] = _snake_to_camel(kwargs[key], True)
+                elif key == "sku":
+                    ev = kwargs[key]
+                    self.parameters["sku"] = Sku(name = ev['name'], tier = ev['tier'], size = ev['size'], family = ev['family'], capacity = ev['capacity'] )
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_sqldatabase.py]:
```suggestion
                    self.parameters["sku"] = Sku(name=ev['name'], tier=ev['tier'], size=ev['size'], family=ev['family'], capacity=ev['capacity'])
```

File: plugins/modules/azure_rm_sqldatabase.py [hunk 3]
```diff
@@ -383,9 +393,9 @@ def exec_module(self, **kwargs):
                 if (('max_size_bytes' in self.parameters) and
                         (self.parameters['max_size_bytes'] != old_response['max_size_bytes'])):
                     self.to_do = Actions.Update
-                if (('edition' in self.parameters) and
-                        (self.parameters['edition'] != old_response['edition'])):
-                    self.to_do = Actions.Update
+                if (('sku' in self.parameters) and
+                         (self.parameters['sku'] != old_response['sku'])):
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_sqldatabase.py]: continuation line over-indented for visual indent

File: plugins/modules/azure_rm_sqldatabase.py [hunk 4]
```diff
@@ -383,9 +393,9 @@ def exec_module(self, **kwargs):
...
+                     self.to_do = Actions.Update
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_sqldatabase.py]: indentation is not a multiple of four and over-indented!

File: plugins/modules/azure_rm_sqldatabase.py (hunk 3 above)
Comment (guyomog78) [reply plugins/modules/azure_rm_sqldatabase.py]: I use the same style usage already present. should I update my part?

File: plugins/modules/azure_rm_sqldatabase.py (hunk 4 above)
Comment (guyomog78) [reply plugins/modules/azure_rm_sqldatabase.py]: Ouch, shame on me! I correct It right now

File: plugins/modules/azure_rm_sqldatabase.py (hunk 1 above)
Comment (guyomog78) [reply plugins/modules/azure_rm_sqldatabase.py:96]: I have added missing arguments in the documentation

**Comment by Fred-sun** [comment]
@guyomog78 Can you help tests/integration/the targets/azure_rm_sqlserver/tasks/main yml increase test cases? Thank you very much!

**Comment by guyomog78** [comment]
@Fred-sun Like make a test with the S0 mode and ensure that realy is a S0 mode, then change to P1 mode ensure that is well a P1?? someting like that?

**Comment by Fred-sun** [comment]
> @Fred-sun Like make a test with the S0 mode and ensure that realy is a S0 mode, then change to P1 mode ensure that is well a P1?? someting like that?

yes! Thanks!

**Comment by guyomog78** [comment]
OK, did you have an ETA in mind?

**Comment by guyomog78** [comment]
@Fred-sun, it was a little bit more difficulte as expected, but necessary to correct the sqldatabasinfo too to report the good SKU like documentation say.
when database information collected, sku name is equal to Standard for Service_objective_name S0,S1,S2.. Premium for Service_objective_name P1, P2 ,P3 ...
So, to use the expected information in sku.name (ex: S0, P2, DW100c , GP_FsV2_12...) I have used the current_service_objective_name retruned by the request.

To see current information: `Get-AzSqlServerServiceObjective -Location eastus`

**Comment by Fred-sun** [comment]
@guyomog78 Such parameter changes may cause the edition parameter previously configured by the user to be unavailable. Can the following two methods be used to avoid this problem?

First: Add Fail prompt. After this version, the edition parameter is changed to sku, and sku aliase is changed to sku.

Second: Keep the edition parameter, but set the default value and "mutually_exclusive=['edition', 'sku']", for example, when edition: Standard, the default sku.name=S1, and edition: premium, the default sku.name=P1.

**Comment by guyomog78** [comment]
@Fred-sun t it is a while that the edition parameter is not functionnal, but I understand your point of view.

I will try to keep for this time the edition parameters

**Comment by guyomog78** [comment]
@Fred-sun It's Done!!
I have also used autopep8 and pylint in vscode and applied "format document"

On test file, I have injected Edition value Premium and it's good. I have tested with standard too ;-)

et Voilà!

PS: Feel free to update the version where edition will be full removed

File: plugins/modules/azure_rm_sqldatabase.py [hunk 5]
```diff
@@ -298,6 +365,7 @@ def __init__(self):
                 choices=['present', 'absent']
             )
         )
+        mutually_exclusive = [['sky', 'edition']]
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_sqldatabase.py]:
```suggestion
        mutually_exclusive = ['sku', 'edition']
```

File: plugins/modules/azure_rm_sqldatabase.py [hunk 6]
```diff
@@ -5,6 +5,8 @@
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
 from __future__ import absolute_import, division, print_function
+from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase, format_resource_id
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_sqldatabase.py]: Import found before documentation variables

File: plugins/modules/azure_rm_sqldatabase.py [hunk 7]
```diff
@@ -5,6 +5,8 @@
...
+import time
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_sqldatabase.py]: Import found before documentation variables

File: plugins/modules/azure_rm_sqldatabase.py (hunk 5 above)
Comment (guyomog78) [reply plugins/modules/azure_rm_sqldatabase.py]: Oups!. done

**Comment by guyomog78** [comment]
@Fred-sun Thanks a lot for review.
Now I'm aware about that the import part for script should be after documentation. It was modified by the "format document" under vscode. Now, I know that I need to revert this part for the next time if appens :-)

**Pull request merged**